### PR TITLE
Port to wx python 4.2.1 and Python3

### PR DESCRIPTION
**Pull request by chrisinabox**
Updates the OD Edit tool to use the latest wxPython version.  The UI now works with Python 3.
This change does not do any significant refactoring to improve the underlying logic.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,24 +51,6 @@ the top in a command-prompt (here assuming Windows and git bash):
 After this `venv/Scripts/odg.exe` will exist and can be called
 from anywhere to run it.
 
-
```
Comment (sveinse) [inline README.md:55]: I don't think this PR should contain any py2.7 removal. Let's do that in a separate cleanup-commit.

File: src/objdictgen/node.py [hunk 2]
```diff
@@ -79,14 +80,39 @@ def StringFormat(text, idx, sub):  # pylint: disable=unused-argument
     if result:
         fmt = result.groups()
         try:
-            log.debug("EVAL StringFormat(): '%s'" % (fmt[1],))
-            return fmt[0] % eval(fmt[1])  # FIXME: Using eval is not safe
+            args = fmt[1].split(',')
```
Comment (sveinse) [inline src/objdictgen/node.py:83]: While it's great to get rid of evaluate (we need that), this part belongs in a separate PR.

File: src/objdictgen/node.py [hunk 3]
```diff
@@ -53,7 +54,7 @@
 Fore = colorama.Fore
 Style = colorama.Style
 
-RE_NAME = re.compile(r'(.*)\[(.*)\]')
+RE_NAME = re.compile(r'(.*)\[[(](.*)[)]\]')
```
Comment (sveinse) [inline src/objdictgen/node.py:57]: Why this change? Please add a comment for this. (But in another PR, please).

File: src/objdictgen/node.py [hunk 4]
```diff
@@ -984,7 +1010,7 @@ def GetTypeList(self):
         list_ = Find.TypeList(MAPPING_DICTIONARY)
         for mapping in self.GetMappings():
             list_.extend(Find.TypeList(mapping))
-        return ",".join(sorted(list_))
+        return list_
```
Comment (sveinse) [inline src/objdictgen/node.py:1013]: Why can we use true lists here instead of comma separated string? Is it because of new wx UI handling?

File: src/objdictgen/ui/subindextable.py [hunk 5]
```diff
@@ -39,16 +39,12 @@
 
 RW = ["Read Only", "Write Only", "Read/Write"]
 RO = ["Read Only", "Read/Write"]
-RW_ACCESS_LIST = ",".join(RW)
```
Comment (sveinse) [inline src/objdictgen/ui/subindextable.py:42]: Are these removals also due to wx no longer needing the these vars to be comma separated strings?

File: tests/test_node.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import pytest
```
Comment (sveinse) [inline tests/test_node.py:1]: This is excellent. Test cases for real-world strings should also be included here I think. E.g. like `$NODEID+12` (random example) or some such.

**Review by sveinse — CHANGES_REQUESTED** [review]
This PR is a lon awaited contribution. Thank you very much @chrisinabox . I really appreciate the efforts and it will be a great addition to the tool. However, we're not quite there yet and some changes are needed.

AFAICS there are three changes included in this PR:
* wx support for py3
* The removal of py2 (at least in the comments and setup)
* Improvements on string parsing

These three belongs in separate PRs. Please adjust the PR accordingly. I recommend making this the wx change and make the two other changes later. If the string parsing is a prerequisite for the wx change, I recommend filing a new PR for that and that it will be approved before this.

The python package code check failed. It does not seem to be able to install wx without compilation. Please check the logs. Getting this working is also something that must be fixed.

**Comment by chrisinabox** [comment]
> This PR is a long awaited contribution. Thank you very much @chrisinabox . I really appreciate the efforts and it will be a great addition to the tool. However, we're not quite there yet and some changes are needed.
> 
> AFAICS there are three changes included in this PR:
> 
> * wx support for py3
> * The removal of py2 (at least in the comments and setup)
> * Improvements on string parsing
> 
> These three belongs in separate PRs. Please adjust the PR accordingly. I recommend making this the wx change and make the two other changes later. If the string parsing is a prerequisite for the wx change, I recommend filing a new PR for that and that it will be approved before this.
> 
> The python package code check failed. It does not seem to be able to install wx without compilation. Please check the logs. Getting this working is also something that must be fixed.

I believe your summary is correct on the most part. In addition though I have also implemented a couple of small bug fixes to the UI which are now part of this pull request. These changes are:

- Allowing deletion of record types objects from the manufacturer specific information section for a node
- Remove a bug that added "ghost" data to the UserMapping dictionary when adding manufacturer info to the node

File: tests/test_node.py (hunk 6 above)
Comment (chrisinabox) [reply tests/test_node.py:1]: I added these tests to allow testing of the methods I introduced to remove the use of eval() which as you suggest above should be part of another PR. This is a very very basic expression parser which allows addition and subtraction. The cases the functions support are based on the existing variable names in the maps.py. i.e.:

`"name": "Restore Manufacturer Defined Default Parameters %d[(sub - 3)]"` (line 184, maps.py)

File: src/objdictgen/node.py (hunk 3 above)
Comment (chrisinabox) [reply src/objdictgen/node.py:57]: This regex expression is used to match variable names from maps.py. It is only used in the StringFormat(..) method and is used to separate the unformatted string from the parameters to be used in the formatting. When I removed the use of eval() I adjusted the regex to match and find the parameters within the brackets, rather that including the round brackets. I then treat this as a comma separated list rather than using it directly in the string formatting with the eval() method.

This way of formatting the names in now partially baked into the .jsonc implementation. See an example from a new OD below:

`"index": "0x1A00",  // 6656
      "name": "Transmit PDO %d Mapping[(idx)]",
      "struct": "narray",
      "group": "built-in",
      "mandatory": false,
      "incr": 1,
      "nbmax": 512,
      "each": {
        "name": "PDO %d Mapping for a process data variable %d[(idx,sub)]",
        "type": "UNSIGNED32",  // 7
        "access": "rw",
        "pdo": false,
        "nbmin": 0,
        "nbmax": 64
      },`

File: src/objdictgen/node.py (hunk 4 above)
Comment (chrisinabox) [reply src/objdictgen/node.py:1013]: Correct, the latest implementation of GridCellChoiceEditor expects a true list -> https://docs.wxpython.org/wx.grid.GridCellChoiceEditor.html?highlight=gridcellchoiceeditor#wx.grid.GridCellChoiceEditor

From my investigation into the previous implementation this method is exclusively used when generating the UI drop downs and therefore I thought it appropriate to modify the base functions in node.py.

File: src/objdictgen/ui/subindextable.py (hunk 5 above)
Comment (chrisinabox) [reply src/objdictgen/ui/subindextable.py:42]: Yes, see comment here https://github.com/Laerdal/python-objdictgen/pull/2#discussion_r1405524995

These cs strings were only used in the GridCellChoiceEditor constructors and I therefore removed them.

File: README.md (hunk 1 above)
Comment (chrisinabox) [reply README.md]: Commit has been reverted (https://github.com/Laerdal/python-objdictgen/pull/2/commits/8afaef5193153b0469732214d348bace8c6f546b) and cherry-picked onto separate branch (https://github.com/chrisinabox/python-objdictgen/tree/cleanup-remove-references-to-py2)

File: src/objdictgen/node.py (hunk 2 above)
Comment (chrisinabox) [reply src/objdictgen/node.py]: I have partially reverted the commit with these changes (https://github.com/Laerdal/python-objdictgen/pull/2/commits/9168780cb4e4de4881f4b2edf8e01b83e142ee4c).

I will move these changes to a separate branch/PR as requested.

File: src/objdictgen/node.py (hunk 3 above)
Comment (chrisinabox) [reply src/objdictgen/node.py]: Change has now been removed from this PR.

**Comment by sveinse** [comment]
The check passed with you change 👍. However it took 23 minutes (!) where compiling wx took 22 of them. I think we need to pursue a better ways to do this. The problem is not that it takes CPU time, but its building something that isn't ours. Need to look into if there is a way to cache wx.

File: src/objdictgen/nodemanager.py [hunk 7]
```diff
@@ -1047,6 +1047,9 @@ def GetNodeEntryValues(self, node, index):
                                 except ValueError as exc:
                                     log.debug("ValueError: '%s': %s" % (values[1], exc))
                                     raise  # FIXME: Originial code swallows exception
+                                except TypeError as exc:
+                                    log.debug("ValueError: '%s': %s" % (dic["value"], exc))
+                                    pass # FIXME: dict["value"] can contain $NODEID for PDOs which is not an int i.e. $NODEID+0x200
```
[the remark this answers is not in the document]
Comment (chrisinabox) [reply src/objdictgen/nodemanager.py:1052]: Valid point. This was quite a hacky fix to prevent a UI crash, unfortunately the previous implementation seemed to expect to fail.

[the remark this answers is not in the document]
Comment (sveinse) [reply src/objdictgen/nodemanager.py:1052]: There is a number of places in this code base that absorbs errors such as this. That's why I had them tagged with `FIXME`'s. Silently ignored errors must be handled deliberately and with care.

**Comment by sveinse** [comment]
I've run the PR on windows and briefly tested it. It works out of box. 🥇 great job. So I'm giving a green light to the PR ✔️ .

Please make the two following changes before merging:
- Figure out what you want to do on the two except statement comment above
- Bump `odg` to version 3.4 (search for 3.3 and replace with 3.4)

**Comment by chrisinabox** [comment]
> I've run the PR on windows and briefly tested it. It works out of box. 🥇 great job. So I'm giving a green light to the PR ✔️ .
> 
> Please make the two following changes before merging:
> 
> * Figure out what you want to do on the two except statement comment above
> * Bump `odg` to version 3.4 (search for 3.3 and replace with 3.4)

Shall get these done ASAP.

File: src/objdictgen/nodemanager.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisinabox) [reply src/objdictgen/nodemanager.py:1053]: This has now fixed in [the latest commit.](https://github.com/Laerdal/python-objdictgen/pull/2/commits/4693e72390f270a406decb3ebdc4295ced6fba5f)

**Review by sveinse — APPROVED** [review]
